config/src/generator.rs

```diff
@@ -0,0 +1,185 @@
+
+/// Produces a new set of FullNodes that connect to the specified upstream peer. The resulting
+/// configs copy all relevant data from this peer including the data used for generating the
+/// ValidatorVerifier. It will also copy the genesis.blob if it is Some.
+pub fn full_node_swarm(
+    upstream_peer: &mut NodeConfig,
+    mut template: NodeConfig,
+    num_nodes: usize,
+    prune_seed_peers: bool,
+    is_ipv4: bool,
+    key_seed: Option<[u8; 32]>,
+    is_permissioned: bool,
+) -> Result<Vec<NodeConfig>> {
+    init(&mut template, is_ipv4);
+    let seed = key_seed.unwrap_or([0u8; 32]);
+    let mut rng = StdRng::from_seed(seed);
+
+    let mut upstream_network = if let Some(network) = template.full_node_networks.get(0) {
+        network.clone_for_template()
+    } else {
+        NetworkConfig::default()
+    };
+    upstream_network.is_permissioned = is_permissioned;
+    upstream_network.random(&mut rng);
+
+    upstream_network.listen_address = utils::get_available_port_in_multiaddr(is_ipv4);
+    upstream_network.advertised_address = upstream_network.listen_address.clone();
+
+    let mut network_peers = NetworkPeersConfig {
+        peers: HashMap::new(),
+    };
+    let mut seed_peers = SeedPeersConfig {
+        seed_peers: HashMap::new(),
+    };
+    add_peer(&upstream_network, &mut network_peers, &mut seed_peers);
+
+    template.set_role(RoleType::FullNode)?;
+    template.consensus.consensus_peers = upstream_peer.consensus.consensus_peers.clone();
+    template.execution.genesis = upstream_peer.execution.genesis.clone();
+    template.state_sync.upstream_peers = UpstreamPeersConfig {
+        upstream_peers: vec![upstream_network.peer_id],
+    };
+
+    let mut configs = Vec::new();
+
+    for _index in 0..num_nodes {
+        let mut config = NodeConfig::random_with_template(&template, &mut rng);
+        config.randomize_ports();
+        let network = &mut config.full_node_networks[0];
+        network.listen_address = utils::get_available_port_in_multiaddr(is_ipv4);
+        network.advertised_address = network.listen_address.clone();
+        add_peer(&network, &mut network_peers, &mut seed_peers);
+        configs.push(config);
+    }
+
+    if prune_seed_peers {
+        seed_peers.seed_peers = seed_peers.seed_peers.into_iter().take(1).collect();
+    }
+
+    for config in &mut configs {
+        config.full_node_networks[0].network_peers = network_peers.clone();
+        config.full_node_networks[0].seed_peers = seed_peers.clone();
+    }
+
+    // Modify upstream peer config to add the new network config.
+    upstream_network.base = upstream_peer.base.clone();
+    upstream_network.network_peers = network_peers;
+    upstream_network.seed_peers = seed_peers;
+    upstream_peer.full_node_networks.push(upstream_network);
+
+    Ok(configs)
+}
+
+/// Generate a set of validator nodes. This does not include the genesis.blob to eliminate crate
+/// dependency issues.
+pub fn validator_swarm(
```

Comment:
Hold on...why do we have two of these?

Reply:
this produces NodeConfigs without genesis
config-builder produces NodeConfigs on disk with genesis

by the looks of things there's an additional optimization that can be made by merging this logic into SwarmConfigBuilder::build

I'll add another commit to provide this.

Reply:
Am I missing something, it doesn't appear as if you ended up merging this in one of the following commits

Reply:
Resolved in last commit.